### PR TITLE
[LLM Evaluate] Adding v1 grading prompt template

### PR DESCRIPTION
**Pull request by sunishsheth2009**
## What changes are proposed in this pull request?
[LLM Evaluate] Adding v1 grading prompt template

## How is this patch tested?
- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/base.py [hunk 1]
```diff
@@ -14,3 +14,35 @@ class MetricValue:
...
+    Stores the sample example during few shot learning during LLM evaluation
+    """
+
+    input: str
+    output: str
+    score: float
+    justification: str = None
+    variables: Dict[str, str] = None
+
+    def toString(self) -> str:
```
Comment (harupy) [inline mlflow/metrics/base.py:31]: can we use `__str__`?

File: mlflow/metrics/base.py [hunk 2]
```diff
@@ -14,3 +14,35 @@ class MetricValue:
...
+        if self.justification is not None:
+            justification = f"Justification: {self.justification}\n"
+
+        return f"""
+        Input: {self.input}
+        Provided output: {self.output}
+        {variables}
+        Score: {self.score}
+        {justification}
+        """
```
Comment (harupy) [inline mlflow/metrics/base.py:48]: /ai remove indent

File: tests/metrics/utils/prompts/test_v1.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        """,
+        examples=[
+            EvaluationExample(
+                input="This is an input",
+                output="This is an output",
+                score=4,
+                justification="This is a justification",
+                variables={"ground_truth": "This is an output"},
+            )
+        ],
```
Comment (prithvikannan) [inline tests/metrics/utils/prompts/test_v1.py:33]: can we add a test case with multiple examples?

File: tests/metrics/test_base.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        )
+    )
+    example1_expected = """
+        Input: This is an input
+        Provided output: This is an output
+        Provided foo: bar
+        Score: 5
+        Justification: This is a justification
+        """
+    assert re.sub(r"\s+", "", example1_expected) == re.sub(r"\s+", "", example1)
```
Comment (prithvikannan) [inline tests/metrics/test_base.py:23]: do we care about the formatting of these strings? if so, should we use `strip()` or maybe not even remove whitespace during these checks.

File: mlflow/metrics/utils/prompts/v1.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                "properties": {
+                    f"reasoning_for_{self.name}": {
+                        "type": "string",
+                        "description": f"Your reasoning for giving the grading for the {self.name} "
+                        "of the output. Provide 5 to 30 words explanation.",
+                    }
+                },
+                "required": [f"reasoning_for_{self.name}"],
+            },
+        }
```
Comment (prithvikannan) [inline mlflow/metrics/utils/prompts/v1.py:94]: im not understanding this part. how is this going to be used? and do we need to build similar versions for pyfunc models.

File: tests/metrics/test_base.py (hunk 4 above)
Comment (sunishsheth2009) [reply tests/metrics/test_base.py:23]: I don't think we care about the exact spacing so this test should ideally be ok if the contents are generated correctly.

File: tests/metrics/utils/prompts/test_v1.py (hunk 3 above)
Comment (sunishsheth2009) [reply tests/metrics/utils/prompts/test_v1.py:33]: Sure. I can add that

File: mlflow/metrics/base.py [hunk 6]
```diff
@@ -14,3 +14,35 @@ class MetricValue:
...
+class EvaluationExample:
+    """
+    Stores the sample example during few shot learning during LLM evaluation
+    """
+
+    input: str
+    output: str
+    score: float
+    justification: str = None
+    variables: Dict[str, str] = None
```
Comment (dbczumar) [inline mlflow/metrics/base.py:29]: Will this be a public / user-facing class (I think it is)? If so, can we add detailed docstrings with examples for each field?

File: mlflow/metrics/utils/prompts/v1.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    Grading criteria:
+    {grading_prompt}
+
+    {examples}
+
+    And you'll need to submit your grading for the {name} of the output,
+    using the following format:
+    Reasoning for {name}: [your step by step reasoning about the {name} of the output]
+    Score for {name}: [your score number between 0 to 4 for the {name} of the output]
```
Comment (dbczumar) [inline mlflow/metrics/utils/prompts/v1.py:44]: Has @suanmiao approved this template?

File: mlflow/metrics/utils/prompts/v1.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+)
+from mlflow.metrics.utils.prompt_template import (
+    PromptTemplate,
+)
+
+default_model = "openai:/gpt4"
+default_parameters = {
+    "temperature": 0.0,
+    "max_tokens": 100,
+    "top_p": 1.0,
```
Comment (dbczumar) [inline mlflow/metrics/utils/prompts/v1.py:15]: Should `top_p` and `temperature` both be specified? @suanmiao my understanding was that you should only set one of these

File: mlflow/metrics/utils/prompts/v1.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            "name": "grading_function",
+            "description": "Call this function to submit the grading for the output",
+            "parameters": {
+                "type": "object",
+                "properties": {
+                    f"reasoning_for_{self.name}": {
+                        "type": "string",
+                        "description": f"Your reasoning for giving the grading for the {self.name} "
+                        "of the output. Provide 5 to 30 words explanation.",
+                    }
```
Comment (dbczumar) [inline mlflow/metrics/utils/prompts/v1.py:90]: @suanmiao does this look good to you / have we tested it?

File: mlflow/metrics/utils/prompts/v1.py (hunk 5 above)
Comment (dbczumar) [reply mlflow/metrics/utils/prompts/v1.py:94]: @sunishsheth2009 Can we make the `grading_function` part of the `EvaluationModel` arguments? Then, we can construct `EvaluationModel` instances with OpenAI grading functions, etc.

> For pyfunc we assume the users would massage the data to make the LLM returns back a particular format.

What is the expected format?

File: mlflow/metrics/utils/prompts/v1.py (hunk 9 above)
Comment (dbczumar) [inline mlflow/metrics/utils/prompts/v1.py:90]: Where does this format come from? This is using OpenAI functions, right? If so, Gateway isn't compatible with OpenAI functions, so how will this work for Gateway judges, e.g. OpenAI thru Gateway?

File: mlflow/metrics/utils/prompts/v1.py (hunk 7 above)
Comment (sunishsheth2009) [reply mlflow/metrics/utils/prompts/v1.py:44]: I am using the https://github.com/databricks/doc_qa/blob/b0ecd4486b784ecbb6047e9bedb2e4a545ed5cbc/databricks/labs/doc_qa/variables/doc_qa_template_variables.py template from @suanmiao

File: mlflow/metrics/utils/prompts/v1.py (hunk 8 above)
Comment (sunishsheth2009) [reply mlflow/metrics/utils/prompts/v1.py:15]: I think we do plan to update this post testing. Will add a TODO to update this

File: mlflow/metrics/utils/prompts/v1.py (hunk 9 above)
Comment (sunishsheth2009) [reply mlflow/metrics/utils/prompts/v1.py:90]: This comes from here as well: https://github.com/databricks/doc_qa/blob/b0ecd4486b784ecbb6047e9bedb2e4a545ed5cbc/databricks/labs/doc_qa/variables/doc_qa_template_variables.py

Comment (sunishsheth2009) [reply mlflow/metrics/utils/prompts/v1.py]: Discussed offline. Removing this for now to make it generic

File: mlflow/metrics/utils/prompts/v1.py (hunk 5 above)
Comment (sunishsheth2009) [reply mlflow/metrics/utils/prompts/v1.py]: Discussed offline. Removing this for now to make it generic

File: mlflow/metrics/utils/prompts/v1.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+# TODO: Update the default_mode and default_parameters to the correct values post experimentation
+default_model = "openai:/gpt4"
+default_parameters = {
+    "temperature": 0.0,
+    "max_tokens": 100,
+    "top_p": 1.0,
+}
+grading_system_prompt_template = PromptTemplate(
+    """
+    Please act as an impartial judge and evaluate the quality of the provided output which
```
Comment (harupy) [inline mlflow/metrics/utils/prompts/v1.py:20]: Can we remove indent?

Comment (harupy) [reply mlflow/metrics/utils/prompts/v1.py:20]: Screen Shot 2023-09-02 at 13 01 31

Screen Shot 2023-09-02 at 13 01 54

File: mlflow/metrics/utils/prompts/v1.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                name=self.name,
+                definition=self.definition,
+                grading_prompt=self.grading_prompt,
+                examples=examples_str,
+            ),
+            "parameters": self.parameters,
+        }
+
+    def _format_examples(self):
+        return "\n".join([str(item) for item in self.examples])
```
Comment (harupy) [inline mlflow/metrics/utils/prompts/v1.py:77]: /ai simplify this function using `map`

**Review by prithvikannan — APPROVED** [review]
lgtm

File: mlflow/metrics/utils/prompts/v1.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+Grading criteria:
+{grading_prompt}
+
+{examples}
+
+And you'll need to submit your grading for the {name} of the output,
+using the following in json format:
+Reasoning for {name}: [your step by step reasoning about the {name} of the output]
+Score for {name}: [your score number between 0 to 4 for the {name} of the output]
```
Comment (dbczumar) [inline mlflow/metrics/utils/prompts/v1.py:45]: @sunishsheth2009 I think we need to specifically tell the LLM what names to use for keys  in the JSON output. If we aren't specific, the LLM may choose different key names, e.g. "reasoning_for_relevance" vs "reasoning for relevance" vs "Reasoning for relevance"

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/9500/files#r1316294707 is addressed. Thanks @sunishsheth2009 !

**Pull request merged**